### PR TITLE
V2

### PR DESCRIPTION
**Pull request by cnuernber**
* Move to new FFI architecture 
* Removed a ton of code
* Simplified implementation
* Refactored to break up implementation into smaller pieces

### CODE REVIEW

File: src/libpython_clj2/python/dechunk_map.clj [hunk 1]
```diff
@@ -0,0 +1,9 @@
+(ns libpython-clj2.python.dechunk-map)
+
+
+(defn dechunk-map
+  "Map a function across a sequence without chunking."
+  [f s]
+  (lazy-seq
+   (when-let [[x] (seq s)]
+     (cons (f x) (dechunk-map f (rest s))))))
```
Comment (jjtolton) [inline src/libpython_clj2/python/dechunk_map.clj:9]: is the cache behavior of `lazy-seq` desirable or undesirable here?

**Comment by jjtolton** [comment]
Whoaaa huge overhaul!  Love it.  Especially love the data oriented touches on `libpython-clj2.python.ffi`.

File: src/libpython_clj2/python/dechunk_map.clj (hunk 1 above)
Comment (cnuernber) [reply src/libpython_clj2/python/dechunk_map.clj:9]: I think it should work the same as map.  This is used specifically in one place where caching the result is ideal and I expected that behavior.

As far as desirable...pluses and minuses :-).

**Comment by cnuernber** [comment]
Huge overhaul.  I am putting lots of work into making sure that there is far less ephemeral python garbage - so for instance before when you called a function (pyfn 1 2 3 4) the python integers would be linked to the GC in order to get cleaned up eventually - now they are cleaned up just after the function call with no GC intervention.  This pathway is much quieter in terms of addref/release logging and much faster than before.  This should also lead to considerably more stable behavior as you won't get large batches of objects getting cleaned up at random times.

And getting to the nitty gritty, to call any python function you have to make a tuple of arguments and the python C pathway to set a tuple item steals the reference so if, for instance, you have an integer and you call PyLong_FromLongLong and set that in the tuple you yourself don't have to decrement the reference created; decrefing the tuple will decref the long value.  

Little details like above lead to much smoother and predictable program behavior and I am really looking forward to all of this getting stabilized and for people to start working with the system.

**Pull request merged**